x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java

```diff
@@ -530,20 +555,48 @@ public boolean shouldCancelChildrenOnCancellation() {
         }
 
         public void stop(String reason, TimeValue timeout) {
-            if (datafeedRunner != null) {
-                datafeedRunner.stopDatafeed(this, reason, timeout);
+            synchronized (this) {
+                if (datafeedRunner == null) {
+                    stoppedOrIsolatedBeforeRunning = StoppedOrIsolatedBeforeRunning.STOPPED;
+                    return;
+                }
             }
+            datafeedRunner.stopDatafeed(this, reason, timeout);
```

Comment:
There are still race conditions inside of `datafeedRunner`

Assume `datafeedRunner.run` has been called, but `runningDatafeedsOnThisNode.put(task.getAllocationId(), holder);` has not yet been set.

If so, during this window between `datafeedRunner.run` and the holder being set, calling `stop` does not prevent the `holder` from being created, added to the table of running datafeeds, and then the task state being set to opened.

Basically `finishHandler` is never called from what I can tell if this occurs. We should do something similar to what we do with the Job task by having a boolean set when `stop` is called that way the thing that creates/sets the holder can stop the datafeed.

Reply:
I think https://github.com/elastic/elasticsearch/pull/75188/commits/3fca44a2003207e1f579ddb4d345ac13f9db0be0 should fix this.

Unlike the job there's no native process to kill so it's not quite a as complicated.  The datafeed doesn't start doing stuff until its persistent task successfully goes into the `started` state, and that shouldn't happen if the task has been removed, unassigned, or its allocation ID has changed.  If any of those things happens then the request to update the datafeed state to `started` will fail with a `ResourceNotFoundException`, and we can remove the entry from the `runningDatafeedsOnThisNode` map (which we should have been doing before in the code that reacts to `ResourceNotFoundException` - it was being leaked before).  Then everything related to the datafeed should be garbage-collected.

I have also added a couple of short-circuits to bail out of the creation process at an earlier point if we detect the datafeed has been stopped or isolated before it gets inserted into the map.